Github repo: https://github.com/<redacted>/assignment-4-equation-of-a-slime-<redacted>

Commit ID: b762a3a3fca826d0264a583a87b63d615c6e7a2c

In [42]:
# Imports section
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression
from sklearn.model_selection import cross_val_score
from sklearn.preprocessing import PolynomialFeatures
from sklearn.metrics import mean_squared_error, r2_score

In [43]:
# Using pandas load the dataset (load remotely, not locally)
# Output the first 15 rows of the data
# Display a summary of the table information (number of datapoints, etc.)
df = pd.read_csv('science_data_large.csv')
print(df.head(15))
print(df.info())

    Temperature °C  Mols KCL     Size nm^3
0              469       647  6.244743e+05
1              403       694  5.779610e+05
2              302       975  6.196847e+05
3              779       916  1.460449e+06
4              901        18  4.325726e+04
5              545       637  7.124634e+05
6              660       519  7.006960e+05
7              143       869  2.718260e+05
8               89       461  8.919803e+04
9              294       776  4.770210e+05
10             991       117  2.441771e+05
11             307       781  5.006455e+05
12             206        70  3.145200e+04
13             437       599  5.390215e+05
14             566        75  9.185271e+04
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 3 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   Temperature °C  1000 non-null   int64  
 1   Mols KCL        1000 non-null   int64  
 2   Size nm^3       1000 n

## Part 2. Splitting the dataset

In [44]:
# Take the pandas dataset and split it into our features (X) and label (y)
X = df.iloc[:, :-1]
y = df.iloc[:, -1]
# Use sklearn to split the features and labels into a training/test set. (90% train, 10% test)
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.1, random_state=42)

## Part 3. Perform a Linear Regression

In [45]:
# Use sklearn to train a model on the training set
model = LinearRegression()
model.fit(X_train, y_train)

# Create a sample datapoint and predict the output of that sample with the trained model
sample_data = pd.DataFrame([[523, 764]], columns=['Temperature °C', 'Mols KCL'])
sample_prediction = model.predict(sample_data)
print(sample_prediction)

# Report on the score for that model, in your own words (markdown, not code) explain what the score means
score = model.score(X_test, y_test)
print(score)

# Extract the coefficents and intercept from the model and write an equation for your h(x) using LaTeX
coefficients = model.coef_
intercept = model.intercept_
print(coefficients, intercept)

[832582.12540526]
0.8552472077276095
[ 866.14641337 1032.69506649] -409391.47958340764


The score explains how well the model explains the variance in the data. If the score is 1 then the model can perfectly predict a variable. If it is 0 then the model cannot explain any varaiablity so it will just give the mean. If it is negative then the model is completely inaccurate to the point it gives a worse result than the mean. In this case we get a 0.8552472077276095 which is inbetween 0 and 1, so it can explain around 85% of the variablity of the model.

Sample equation: $E = mc^2$

$h(x) = -409391.47958340764 + (866.14641337)T + (1032.69506649)Mols$

## Part 4. Use Cross Validation

In [46]:
# Use the cross_val_score function to repeat your experiment across many shuffles of the data
scores = cross_val_score(model, X, y, cv=20)

# Report on their finding and their significance
print(scores)
meanScore = sum(scores)/ 20

standardDeviationScore = (sum(((i - meanScore)**2) for i in scores)/20)**0.5

print(meanScore, standardDeviationScore)


[0.75200707 0.86394464 0.87300135 0.85233434 0.82161172 0.89379119
 0.86923658 0.85923885 0.89198098 0.85767661 0.80900828 0.86227826
 0.88267625 0.87499542 0.87696169 0.84666045 0.76802123 0.80459254
 0.87900272 0.89513303]
0.8517076599630038 0.03949413180642212


The mean score I got was 85.17% variabilty and the scores had a very small standard deviation of 3.95%. So our current model can pretty accurately predict 85% of the variabilty of the model.

## Part 5. Using Polynomial Regression

In [52]:
# Using the PolynomialFeatures library perform another regression on an augmented dataset of degree 2
poly = PolynomialFeatures(degree=2)
X_poly = poly.fit_transform(X)

X_train, X_test, y_train, y_test = train_test_split(X_poly, y, test_size=0.1, random_state=42)

model_poly = LinearRegression()
model_poly.fit(X_train, y_train)

poly_score = model_poly.score(X_test, y_test)
print(poly_score)

coefficients = model_poly.coef_
intercept = model_poly.intercept_

# Report on the metrics and output the resultant equation as you did in Part 3.
print(coefficients, intercept)


1.0
[ 0.00000000e+00  1.20000000e+01 -1.27195488e-07  1.26494371e-11
  2.00000000e+00  2.85714287e-02] 2.0477105863392353e-05


$h(x) = 12T + 2.000 * T*Mols + 0.029Mols^2$

The score is 1 so the model must be perfect